pkg/query/statistics.go

```diff
@@ -0,0 +1,210 @@
+	// This represents the computational cost of verifying if a specific resource-subject
+	// relationship exists.
+	CheckCost int
+
+	// IterResourcesCost is the estimated cost to iterate over all resources
+	// accessible through this iterator.
+	IterResourcesCost int
+
+	// IterSubjectsCost is the estimated cost to iterate over all subjects
+	// that have access through this iterator.
+	IterSubjectsCost int
+}
+
+// StatisticsSource provides cost estimates for iterators.
+// Implementations can provide static estimates or dynamic estimates based on
+// actual datastore statistics.
+type StatisticsSource interface {
+	// Cost returns a cost estimate for the given iterator.
+	Cost(it Iterator) (Estimate, error)
+}
+
+// StaticStatistics provides static cost estimates for iterators based on
+// configurable parameters. This is useful for basic query planning and
+// when dynamic statistics are not available.
+type StaticStatistics struct {
+	// DatastoreSize is the assumed number of tuples in a relation.
+	DatastoreSize int
+
+	// Fanout is the assumed average number of subjects per resource or
+	// resources per subject.
+	Fanout int
+
+	// CheckSelectivity is the default probability (0.0-1.0) that a Check
+	// operation will return true.
+	CheckSelectivity float64
+
+	// IntersectionArrowReduction is an additional reduction factor for
+	// IntersectionArrow selectivity to account for "all subjects" semantics.
+	IntersectionArrowReduction float64
+}
+
+// DefaultStaticStatistics returns a StaticStatistics instance with default values
+func DefaultStaticStatistics() StaticStatistics {
+	return StaticStatistics{
+		DatastoreSize:              10,
+		Fanout:                     5,
+		CheckSelectivity:           0.9,
+		IntersectionArrowReduction: 0.5,
+	}
+}
+
+// Cost returns a cost estimate for the given iterator using static assumptions.
+// It recursively estimates costs for composite iterators by combining the costs
+// of their subiterators according to their operational semantics.
+func (s StaticStatistics) Cost(iterator Iterator) (Estimate, error) {
+	switch it := iterator.(type) {
+	case *FixedIterator:
+		return Estimate{
+			Cardinality:       len(it.paths),
+			CheckCost:         10,
+			CheckSelectivity:  s.CheckSelectivity,
+			IterResourcesCost: len(it.paths),
+			IterSubjectsCost:  len(it.paths),
+		}, nil
+	case *RelationIterator:
+		return Estimate{
+			Cardinality:       s.DatastoreSize,
+			CheckCost:         s.DatastoreSize,
+			CheckSelectivity:  s.CheckSelectivity,
+			IterResourcesCost: s.DatastoreSize * s.Fanout,
+			IterSubjectsCost:  s.DatastoreSize * s.Fanout,
+		}, nil
+	case *Arrow:
+		ls, err := s.Cost(it.left)
+		if err != nil {
+			return Estimate{}, err
+		}
+		rs, err := s.Cost(it.right)
+		if err != nil {
+			return Estimate{}, err
+		}
+		// Arrow does IterSubjects on left (ls.Cardinality iterations),
+		// then Check on right for each subject (rs.CheckCost per iteration)
```

Comment:
Is this always the case, or is this one strategy? What happens if/when this gets rewritten?